File: Webfuel.Api/Api/Report/ReportApi.cs
using MediatR;
using Microsoft.AspNetCore.Mvc;
using Webfuel.Common;
using Webfuel.Domain;
using Webfuel.Reporting;

namespace Webfuel.App
{
    [ApiService]
    [ApiDataSource]
    public static class ReportApi
    {
        public static void RegisterEndpoints(IEndpointRouteBuilder app)
        {
            // Commands

            app.MapPost("api/report", Create)
                .RequireClaim(c => c.CanEditReports);

            app.MapPost("api/report/copy", Copy)
                .RequireClaim(c => c.CanEditReports);

            app.MapPut("api/report", Update)
                .RequireClaim(c => c.CanEditReports);

            app.MapDelete("api/report/{id:guid}", Delete)
                .RequireClaim(c => c.CanEditReports);

            app.MapPost("api/report/run", Run)
                .RequireIdentity();

            app.MapPost("api/report/run-annual", RunAnnualReport)
                .RequireIdentity();

            // Querys

            app.MapPost("api/report/query", Query)
                .RequireIdentity();

            app.MapGet("api/report/{id:guid}", Get)
                .RequireIdentity();

            app.MapGet("api/report/list-head", ListHead)
                .RequireIdentity();
        }

        public static Task<Report> Create([FromBody] CreateReport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<Report> Copy([FromBody] CopyReport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<Report> Update([FromBody] UpdateReport command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task Delete(Guid id, IMediator mediator)
        {
            return mediator.Send(new DeleteReport { Id = id });
        }

        public static Task<QueryResult<Report>> Query([FromBody] QueryReport command, IMediator mediator)
        {
            return mediator.Send(command);
        }
        public static async Task<ReportStep> Run([FromBody] RunReport command, IMediator mediator)
        {
            return await mediator.Send(command);
        }

        public static async Task<ReportStep> RunAnnualReport([FromBody] RunAnnualReport command, IMediator mediator)
        {
            return await mediator.Send(command);
        }

        public static async Task<Report?> Get(Guid id, IMediator mediator)
        {
            return await mediator.Send(new GetReport { Id = id });
        }

        public static async Task<List<Report>> ListHead(IMediator mediator)
        {
            return await mediator.Send(new ListHeadReport());
        }
    }
}

File: Webfuel.App/ClientApp/src/app/api/report.api.ts
import { EventEmitter, Injectable, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService, ApiOptions } from '../core/api.service';
import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { IDataSource } from 'shared/common/data-source';
import { CreateReport, Report, CopyReport, UpdateReport, RunReport, ReportStep, RunAnnualReport, QueryReport, QueryResult } from './api.types';

@Injectable()
export class ReportApi implements IDataSource<Report, QueryReport, CreateReport, UpdateReport> {
    constructor(private apiService: ApiService) { }
    
    public create (body: CreateReport, options?: ApiOptions): Observable<Report> {
        return this.apiService.request<CreateReport, Report>("POST", "api/report", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public copy (body: CopyReport, options?: ApiOptions): Observable<Report> {
        return this.apiService.request<CopyReport, Report>("POST", "api/report/copy", body, options);
    }
    
    public update (body: UpdateReport, options?: ApiOptions): Observable<Report> {
        return this.apiService.request<UpdateReport, Report>("PUT", "api/report", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public delete (params: { id: string }, options?: ApiOptions): Observable<any> {
        return this.apiService.request<undefined, any>("DELETE", "api/report/" + params.id + "", undefined, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public run (body: RunReport, options?: ApiOptions): Observable<ReportStep> {
        return this.apiService.request<RunReport, ReportStep>("POST", "api/report/run", body, options);
    }
    
    public runAnnualReport (body: RunAnnualReport, options?: ApiOptions): Observable<ReportStep> {
        return this.apiService.request<RunAnnualReport, ReportStep>("POST", "api/report/run-annual", body, options);
    }
    
    public query (body: QueryReport, options?: ApiOptions): Observable<QueryResult<Report>> {
        return this.apiService.request<QueryReport, QueryResult<Report>>("POST", "api/report/query", body, options);
    }
    
    public get (params: { id: string }, options?: ApiOptions): Observable<Report> {
        return this.apiService.request<undefined, Report>("GET", "api/report/" + params.id + "", undefined, options);
    }
    
    public listHead (options?: ApiOptions): Observable<Array<Report>> {
        return this.apiService.request<undefined, Array<Report>>("GET", "api/report/list-head", undefined, options);
    }
    
    changed = new EventEmitter<any>();
    
    // Resolvers
    
    static reportResolver(param: string): ResolveFn<Report> {
        return (route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<Report> => {
            return inject(ReportApi).get({id: route.paramMap.get(param)! });
        };
    }
}

